### DISCUSSION TITLE
I added a "controlled crossing" feature

### DISCUSSION THREAD

**Discussion by mgfarmer** [mgth/LittleBigMouse#329] [Show and tell]
Code is on [my fork](https://github.com/mgfarmer/LittleBigMouse/tree/master) (I try to keep this fork synced to upstream)

## Controlled Crossing

_The implementation of this feature is fully functional, but not complete.  See the limitations below.  There is no UI for it._

An installer is available [here](https://github.com/mgfarmer/LittleBigMouse/releases/tag/5.1.0.1-CC).

This feature adds the ability to "control" crossing horizontal and/or vertical edges.  You can configure this feature to prevent crossing the edges unless a control condition is met.  Why would you want this feature?  Well, not everyone will.  I've been using this feature since December '23, and can't imagine working without it now.  I'll describe my configuration and why this is so exceptional (you may, of course, disagree).  

I have 4 monitors.  Three of them are side-by-side and the fourth is mounted above the center row monitor.  The top monitor is primarily used for monitoring team chats and email.  I don't go there often.  The main issue with having a vertically stacked monitor like this is that without a controlled crossing it is too easy to move the mouse between the upper and lower monitor when trying to interact with any UI near the top of the bottom monitor (like main menus in full screen apps, or the RDP control bar which require holding the mouse at the very top pixel of the screen to get it to drop), and any UI near the bottom of the upper monitor (like a taskbar).  This is because I, like probably most people, have spent years simply pushing the mouse to the top or bottom quickly and expecting the mouse to just stop there when needing to get to these UI elements.  With stacked monitors you will just push the mouse way beyond the elements you are aiming for and then have to drag the mouse back again, slowly zeroing in on the element, bouncing between monitors.

![Snag_410d0ca](https://github.com/mgth/LittleBigMouse/assets/86535300/e28e4a1c-7bd7-4cb8-9134-9c3639dd4373)

With the "controlled crossing" feature you can tell LBM to not make the transition from one monitor to the other unless one of two conditions is met:
  - Hold the CTRL key while crossing.  When this is enabled, crossing the edge happens immediately if the CTRL key is pressed.  This works really well for me.  You can even drag windows across the edge while it is pressed.  Drag-and-drop works too. To date I have not run into any limitations using this method.  (There probably are some, but I have encountered none yet).
  - Push the mouse until it crosses.  With this mode enabled, and configured to your liking, you can just continue to push the mouse at the edge you want to cross until it does.  You can configure a "threshold" to control how much pushing is required.

Both modes can be enabled at the same time. 

## Limitations

 - This feature only works in "straight" mode.  It can probably be made to work in corner-crossing mode (CTRL key for sure, not sure about "push" mode), but I don't use that mode.
 - This feature, as implemented today, turns all horizontal edges or all vertical edges into controlled crossings.  You cannot say that only specific edges are controlled.  I started down that path, but it was much more complicated, and I did not really see a need for it.
 - There is no UI to configure this feature.  It is configured using environment variables.  This is mostly because (a) I do not know the Avalonia UI framework at all, and (b) I don't have time to learn it.  (Sorry...)  It would be nice to add UI at some point.
- Since I'm using environment variables, any changes to the configuration require restarting LBM (not just restarting the server).

If none of these environment variables are defined, LBM behaves as it always has.

The 4 environment variables are described below (via copy/paste from the code comments).

## LBM_CC_HORZ_EDGE
    /*
    * LBM_CC_HORZ_EDGE <- Defined or not, value is not used
    *
    * This envar, when defined, makes all horizontal edges into
    * controlled crossings. The primary use case is when you
    * have vertically stacked monitors. For instance, I have 3
    * side-by-side monitors, with a fourth monitor above the middle
    * monitor. Without controlled crossings interacting with the
    * main menu of full screen apps (near the top edge), or interacting 
    * with the taskbar on the upper monitor (near the bottom edge) is 
    * frustrating because the cursor too easily jumps to the monitor 
    * above or below the interaction zone. This is especially true
    * when using full-screen RDP which requires holding the cursor at 
    * the very top of the screen.
    */

## LBM_CC_VERT_EDGE
   /*
   * LBM_CC_VERT_EDGE <- Defined or not, value is not used
   * 
   * This envar, when defined, makes all vertical edges into
   * controlled crossings.  It is implemented for completeness,
   * though I have not come up with a good use case for it.
   */

## LBM_CC_CTRL_KEY
   /*
   * LBM_CC_CTRL_KEY <- Defined or not, value is not used
   * 
   * The envar, when defined, allows crossing a controlled edge
   * immediately if the Ctrl key is pressed while moving the mouse.
   * This can be used in concert with the LBM_CC_THRESHOLD envar.
   */

## LBM_CC_THRESHOLD
    /*
    * LBM_CC_THRESHOLD  <- Integer > 0
    *
    * This threshold controls how much effort is required to push the mouse
    * cursor across a controlled edge.  Try values around 200 to 250.  The
    * value that works best for you depends a lot on how quickly, or 
    * aggressively, you move your mouse so some tuning may be required.
    *
    * The cursor will stall at the edge until you push long enough, then
    * it will cross.
    * 
    * When undefined, or the value is less than 1, no amount of pushing
    * will allow the cursor across.
    */

If you enable controlled crossing for either horizontal  or vertical, but do not
enable either crossing mode, then the code will disable the edges, otherwise
you would not be able to cross the edge at all.

My hope is that at least two people find this useful (I count as the first person).

**Comment by mgth** [comment]
Hello, I did implement your idea in [Implement-crossing-resistance](https://github.com/mgth/LittleBigMouse/tree/Implement-crossing-resistance) I think it's something very helpful, thank you for the work done. I did not implement ctrl key, do you think it's really needed ? For now its only working in straight mode too and the unit is pixels (I think I will change it to mm to be consistent)

**Comment by mgfarmer** [comment]
Great.  I did not explore your branches! My bad.

I do like the CTRL Key trigger. I actually use it the most.  It has a
slightly more natural feel.  I did the CTRL Key code back in December, and
the "resistance" code just last week.  So I'm very used to the CTRL key now.

-Kevin

**Reply by mgth** [reply]
Sorry my english is not so accurate. I did it yesterday inspired by your code so you would not have seen it before.

**Reply by mipskevin** [reply]
Ah, I understand!  I'll try your branch out.  Also, your English is excellent.
